Network/Socket/Shutdown.hs

```diff
@@ -34,3 +40,40 @@ shutdown s stype = void $ withFdSocket s $ \fd ->
 
 foreign import CALLCONV unsafe "shutdown"
   c_shutdown :: CInt -> CInt -> IO CInt
+
+
+-- | Closing a socket gracefully.
+--   This sends TCP FIN and check if TCP FIN is received from the peer.
+--   The second argument is time out to receive TCP FIN in millisecond.
+--   In both normal cases and error cases, socket is deallocated finally.
+--
+--   Since: 3.1.1.0
+gracefulClose :: Socket -> Int -> IO ()
+gracefulClose s tm = (sendRecvFIN `E.finally` close s) `E.catch` ignore
+  where
+    sendRecvFIN = do
+        -- Sending TCP FIN.
+        shutdown s ShutdownSend
+        -- Waiting TCP FIN.
+        E.bracket (mallocBytes bufSize) free recvEOF
+    recvEOF buf = loop 0
+      where
+        loop delay = do
+            -- We don't check the (positive) length.
+            -- In normal case, it's 0. That is, only FIN is received.
+            -- In error cases, data is available. But there is no
+            -- application which can read it. So, let's stop receiving
+            -- to prevent attacks.
+            r <- recvBufNoWait s buf bufSize
+            let delay' = delay + clock
+            when (r == -1 && delay' < tm) $ do
+                threadDelay (clock * 1000)
```

Comment:
Instead of looping with a `threadDelay`, what if we had an `MVar` which can be filled with either a `MoreData` or `TimeoutTripped` data constructor. We use `registerTimeout` to set a timeout to `putMVar` the `TimeoutTripped` constructor, and `registerFd` to `putMVar` the `MoreData`. Then we don't need an arbitrary delay length and avoid any kind of arbitrary delays or busy-looping.

Reply:
@snoyberg Does 5ee9828 implement what you want?

Reply:
@Mistuke GHC for Windows does not provide `GHC.Event`. Should we use the previous busy-loop implementation for Windows?